x/superfluid/keeper/stake_test.go

```diff
@@ -408,6 +408,152 @@ func (suite *KeeperTestSuite) TestSuperfluidUnbondLock() {
+		},
+		{
+			name:            "unlock amount > locked amount",
+			testInvalidLock: false,
+			unlockAmount:    sdk.NewInt(lockAmount + 1),
+			expectErr:       true,
+			splitLockId:     false,
+		},
+		{
+			name:            "lock is not split if unlock amount = locked amount",
+			testInvalidLock: false,
+			unlockAmount:    sdk.NewInt(lockAmount),
+			expectErr:       false,
+			splitLockId:     false,
+		},
+		{
+			name:            "lock is split if unlock amount < locked amount",
+			testInvalidLock: false,
+			unlockAmount:    sdk.NewInt(lockAmount / 2),
+			expectErr:       false,
+			splitLockId:     true,
+		},
+	}
+
+	for _, tc := range testCases {
+		suite.Run(tc.name, func() {
+			suite.SetupTest()
+
+			// setup validators
+			valAddrs := suite.SetupValidators([]stakingtypes.BondStatus{stakingtypes.Bonded})
+
+			denoms, _ := suite.SetupGammPoolsAndSuperfluidAssets([]sdk.Dec{sdk.NewDec(20), sdk.NewDec(20)})
+
+			// setup superfluid delegations
+			_, intermediaryAccs, locks := suite.setupSuperfluidDelegations(valAddrs, []superfluidDelegation{{0, 0, 0, lockAmount}}, denoms)
+			suite.checkIntermediaryAccountDelegations(intermediaryAccs)
+			suite.Require().True(len(locks) > 0)
+
+			// test invalid lock
+			if tc.testInvalidLock {
+				lock := lockuptypes.PeriodLock{}
+				_, err := suite.App.SuperfluidKeeper.SuperfluidUndelegateAndUnbondLock(suite.Ctx, lock.ID, lock.GetOwner(), sdk.NewInt(1))
+				suite.Require().Error(err)
+				return
+			}
+
+			for _, lock := range locks {
+				startTime := time.Now()
+				suite.Ctx = suite.Ctx.WithBlockTime(startTime)
+				accAddr := suite.App.SuperfluidKeeper.GetLockIdIntermediaryAccountConnection(suite.Ctx, lock.ID)
+				intermediaryAcc := suite.App.SuperfluidKeeper.GetIntermediaryAccount(suite.Ctx, accAddr)
+				valAddr := intermediaryAcc.ValAddr
+
+				unbondLockStartTime := startTime.Add(time.Hour)
+				suite.Ctx = suite.Ctx.WithBlockTime(unbondLockStartTime)
+				lockId, err := suite.App.SuperfluidKeeper.SuperfluidUndelegateAndUnbondLock(suite.Ctx, lock.ID, lock.GetOwner(), tc.unlockAmount)
+				if tc.expectErr {
+					suite.Require().Error(err)
+					continue
+				}
+
+				suite.Require().NoError(err)
+
+				if tc.splitLockId {
+					suite.Require().Equal(lockId, lock.ID+1)
+
+					// check original underlying lock
+					updatedLock, err := suite.App.LockupKeeper.GetLockByID(suite.Ctx, lock.ID)
+					suite.Require().NoError(err)
+					suite.Require().False(updatedLock.IsUnlocking())
+					suite.Require().Equal(updatedLock.Coins[0].Amount, lock.Coins[0].Amount.Sub(tc.unlockAmount))
+
+					// check newly created underlying lock
+					newLock, err := suite.App.LockupKeeper.GetLockByID(suite.Ctx, lockId)
+					suite.Require().NoError(err)
+					suite.Require().True(newLock.IsUnlocking())
+					suite.Require().Equal(newLock.Coins[0].Amount, tc.unlockAmount)
+
+					// check original synthetic lock
+					stakingDenom := keeper.StakingSyntheticDenom(lock.Coins[0].Denom, valAddr)
```

Comment:
Can we also add check for unstaking synth denom for `tc.splitLockId` and check for staking synth denom in the `!tc.splitLockId` cases?